godot_project/project_workspace/workspace_context/history/history.gd

```diff
@@ -180,6 +187,18 @@ func _validate_extern_snapshot(in_snapshot: Dictionary) -> bool:
 	return true
 
 
+## Detects if create_snapshot is called multiple times in a single frame
+## External snapshots pushed by `push_and_apply_snapshot(in_snapshot, in_snapshot_name)`
+## are not considered by this check
+func _monitor_redundant_snapshots(in_snapshot_name: String) -> void:
+	if _last_snapshot_taken_at_frame != Engine.get_frames_drawn():
+		return
+	if REDUNDANT_ACTION_WHITELIST.has(in_snapshot_name) and in_snapshot_name != _last_snapshot_name:
```

Comment:
why we need the second part of this condition? `in_snapshot_name != _last_snapshot_name`  
As far as I understand what it will do it will make this error print a bit less helpful, I assume there probably should not be many situations where snapshot is created couple times in the same frame no matter the source of the snapshot (it should always react to user actions and it's uncommon for user to perform multiple actions in the same frame)

Reply:
if i understand correctly, this is because the point of this method is to catch the case where 2 snapshots **of the same type** are executed **in the same frame** so this early return is simply a whitelist of "known snapshots" that will in fact be created in the same frame 🤔

Reply:
That second part is there to limit false negatives.

If we don't check that, then two accidental 'Apply simulation state' in the same frame would not be detected by this method. Even if they are in the redundant whitelist, there's no reason to have two of these actions in the same frame.

Thinking about it, this condition isn't even complete because it makes the snapshot order important. I should also check if the last_snapshot_name is whitelisted or not.